pkg/agent/downloader.go

```diff
@@ -0,0 +1,85 @@
+package agent
+
+import (
+	"fmt"
+	pro "github.com/kubeflow/kfserving/pkg/agent/protocols"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+type Downloader struct {
+	// TODO: Add back-off retries
+	ModelDir   string
+	ProtocolManagers map[pro.Protocol]pro.ProtocolManager
+}
+
+var SupportedProtocols = []pro.Protocol{pro.S3}
+
+func (d *Downloader) DownloadModel(event EventWrapper) error {
+	modelSpec := event.ModelSpec
+	if modelSpec != nil {
+		modelUri := modelSpec.StorageURI
+		hashString := hash(modelUri)
+		log.Println("Processing:", modelUri, "=", hashString)
+		successFile := filepath.Join(d.ModelDir, fmt.Sprintf("SUCCESS.%d", hashString))
```

Comment:
This seems like an odd mechanism for checking completion. Can't you just check the existence of the file you're trying to download?

Reply:
We create a file and then use the downloader to write to it. if the downloader fails, we might have a partially filled writer. i.e. data corruption or whatnot. So I think it is safer to use a success file. but I don't mind